java-connectors/kafka-connect-azure-eventhubs/src/main/java/io/lenses/streamreactor/connect/azure/eventhubs/source/TopicPartitionOffsetProvider.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2017-2024 Lenses.io Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.lenses.streamreactor.connect.azure.eventhubs.source;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.connect.storage.OffsetStorageReader;
+
+/**
+ * This class represents an abstraction over OffsetStorageReader that can be freely called by Azure
+ * EventHub Kafka Consumers when it was initialized once. It helps consumers to find out which
+ * offset were already committed inside Kafka Connect.
+ */
+@Slf4j
+public final class TopicPartitionOffsetProvider {
+
+  private static final String OFFSET_KEY = "OFFSET";
+
+  private final OffsetStorageReader offsetStorageReader;
+
+
+  public TopicPartitionOffsetProvider(OffsetStorageReader offsetStorageReader) {
+    this.offsetStorageReader = offsetStorageReader;
+  }
+
+  /**
+   * Checks for committed offsets for topic+partition combo.
+   *
+   * @param azureTopicPartitionKey key of topic+partition combo.
+   *
+   * @return empty optional if topic+partition combo has not committed any offsets or
+   *     AzureOffsetMarker if combo already did commit some.
+   */
+  public Optional<AzureOffsetMarker> getOffset(AzureTopicPartitionKey azureTopicPartitionKey) {
+    Optional<AzureOffsetMarker> toReturn = Optional.empty();
+    Optional<Map<String, Object>> offsetOptional = Optional.ofNullable(
+        offsetStorageReader.offset(azureTopicPartitionKey));
+
+    if (offsetOptional.isPresent()) {
+      Long offset = (Long) offsetOptional.get().get(OFFSET_KEY);
+      if (offset != null) {
+        toReturn = Optional.of(new AzureOffsetMarker(offset));
+      }
+    }
+
+    return toReturn;
+  }
```

Comment:
You shouldn't ever really need to call `.get()` on an Optional.  `.map` is useful for transforming it.  You could do something like this to go in a more functional style:

```suggestion
  public Optional<AzureOffsetMarker> getOffset(AzureTopicPartitionKey azureTopicPartitionKey) {
    return Optional.ofNullable(offsetStorageReader.offset(azureTopicPartitionKey))
            .map(offsetMap -> (Long) offsetMap.get(OFFSET_KEY))
            .filter(offset -> offset != null)
            .map(AzureOffsetMarker::new);
  }

```